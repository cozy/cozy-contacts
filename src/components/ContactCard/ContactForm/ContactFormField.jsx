import React from 'react'
import PropTypes from 'prop-types'
import { FieldArray } from 'react-final-form-arrays'
import Button from 'cozy-ui/transpiled/react/Button'
import Icon from 'cozy-ui/transpiled/react/Icon'
import palette from 'cozy-ui/transpiled/react/palette'
import uniqueId from 'lodash/uniqueId'
import { Media, Img, Bd } from 'cozy-ui/transpiled/react/Media'

class ContactFormField extends React.Component {
  constructor(props) {
    super(props)
    this.fieldID = uniqueId('field_')
  }

  render() {
    const { name, icon, isArray, renderInput, t } = this.props
    return (
      <Media align="top" className="contact-form-field">
        <Img>
          {icon ? (
            <Icon
              icon={icon}
              color={palette['coolGrey']}
              className="contact-form-field__icon u-mr-2"
            />
          ) : (
            <div className="u-w-1 u-mr-2" />
          )}
        </Img>
        <Bd>
          {isArray ? (
            <FieldArray name={name}>
              {({ fields }) => (
                <div className="u-mt-1 u-mb-half">
                  {fields.map((nameWithIndex, index) =>
                    this.renderArrayField(
                      fields,
                      index,
                      nameWithIndex,
                      name,
                      renderInput
                    )
                  )}
                  <Button
                    icon="plus"
                    theme="text"
                    type="button"
                    onClick={() => this.addField(fields)}
                    label={t(`addLabel.${name}`)}
                    className="u-ph-0 u-mh-0"
                  />
                </div>
              )}
            </FieldArray>
          ) : (
            <div className="u-mt-1">{renderInput(name, this.fieldID)}</div>
          )}
        </Bd>
      </Media>
    )
  }

  addField = fields => fields.push(undefined)

  removeField = (fields, index) => {
    const isLastRemainignField = fields.length === 1
    fields.remove(index)
    if (isLastRemainignField) this.addField(fields)
  }

  renderArrayField = (fields, index, nameWithIndex, name, renderInput) => {
    const hasValue = fields.value[index] && fields.value[index][name]
    return (
      <div key={nameWithIndex} className="u-mt-1 u-flex u-pos-relative">
        {renderInput(`${nameWithIndex}.${name}`, this.fieldID)}

        {hasValue && (
          <Button
            type="button"
            theme="secondary"
            label="delete"
            iconOnly
            round
            icon={'cross-small'}
            onClick={() => this.removeField(fields, index)}
            className="u-pos-absolute u-top-0 u-right-0 u-mt-1"
          />
        )}
      </div>
    )
  }
}
ContactFormField.propTypes = {
  name: PropTypes.string.isRequired,
  icon: PropTypes.any, // shall be a SVG prop type
  isArray: PropTypes.bool,
  renderInput: PropTypes.func.isRequired
}
ContactFormField.defaultProps = {
  icon: null,
  isArray: false
}

export default ContactFormField
