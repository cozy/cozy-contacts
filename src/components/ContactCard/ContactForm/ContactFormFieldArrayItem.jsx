import React from 'react'

import Icon from 'cozy-ui/transpiled/react/Icon'
import IconButton from 'cozy-ui/transpiled/react/IconButton'
import CrossCircleIcon from 'cozy-ui/transpiled/react/Icons/CrossCircle'
import ListItemIcon from 'cozy-ui/transpiled/react/ListItemIcon'

import { removeField } from '../../../helpers/fields'

const ContactFormFieldArrayItem = ({
  fields,
  index,
  nameWithIndex,
  name,
  renderInput
}) => {
  const showRemove = fields.value[index]?.[name]

  return (
    <div className="u-mt-1 u-flex u-pos-relative">
      {renderInput(`${nameWithIndex}.${name}`)}

      {showRemove && (
        <ListItemIcon className="contact-form-field__delete-icon">
          <IconButton
            aria-label="delete"
            color="error"
            size="medium"
            onClick={() => removeField(fields, index)}
          >
            <Icon icon={CrossCircleIcon} />
          </IconButton>
        </ListItemIcon>
      )}
    </div>
  )
}

export default ContactFormFieldArrayItem
