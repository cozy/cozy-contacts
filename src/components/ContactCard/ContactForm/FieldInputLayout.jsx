import cx from 'classnames'
import PropTypes from 'prop-types'
import React from 'react'

import Icon from 'cozy-ui/transpiled/react/Icon'
import { useI18n } from 'cozy-ui/transpiled/react/providers/I18n'

import { fieldsRequired } from './helpers'
import FieldInput from '../FieldInput'
import FieldInputArray from '../FieldInputArray'

const FieldInputLayout = ({
  attributes: { isArray, ...attributes }, // ⚠️ isArray is here removed from attributes to avoid DOM propagration
  contacts,
  formProps
}) => {
  const { t } = useI18n()
  const { valid, submitFailed, errors } = formProps
  const { name, label, icon, ...restAttributes } = attributes

  const isError = fieldsRequired.includes(name) && !valid && submitFailed

  return (
    <div
      className={cx('u-flex u-mt-1', {
        'u-flex-items-center': !isArray,
        'u-flex-items-baseline': isArray
      })}
    >
      <div className="u-w-2-half">
        {icon && <Icon icon={icon} color="var(--iconTextColor)" />}
      </div>
      <div className="u-w-100">
        {isArray ? (
          <FieldInputArray
            attributes={attributes}
            contacts={contacts}
            formProps={formProps}
          />
        ) : (
          <FieldInput
            attributes={restAttributes}
            contacts={contacts}
            error={isError}
            helperText={isError ? errors[name] : null}
            name={name}
            label={t(`fields.${name}`)}
            labelProps={label}
          />
        )}
      </div>
    </div>
  )
}

FieldInputLayout.propTypes = {
  attributes: PropTypes.object,
  contacts: PropTypes.shape({
    data: PropTypes.arrayOf(PropTypes.object)
  }),
  formProps: PropTypes.object
}

export default FieldInputLayout
