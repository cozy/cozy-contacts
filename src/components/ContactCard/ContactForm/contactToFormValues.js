import { fields } from './index'
import { getFormattedAddress } from '../ContactFields'

export const moveToHead = shouldBeHead => items =>
  items.reduce((arr, v) => (shouldBeHead(v) ? [v, ...arr] : [...arr, v]), [])

const movePrimaryToHead = moveToHead(v => v.primary)

const contactToFormValues = (contact, t) => {
  // initialize the form values, required so that array fields start with at least one editable field
  const initialFieldValues = fields.reduce(
    (initialValues, { name, isArray }) => {
      initialValues[name] = isArray ? [undefined] : undefined
      return initialValues
    },
    {}
  )

  if (contact) {
    const {
      address,
      birthday,
      company,
      cozy,
      email,
      name,
      note,
      phone
    } = contact

    const addressValue =
      address && address.length > 0
        ? movePrimaryToHead(address).map(({ type, ...addressInfo }) => ({
            address: getFormattedAddress(addressInfo, t),
            addressLabel: type
          }))
        : [undefined]
    const cozyValue = cozy && cozy.length > 0 ? cozy[0].url : undefined
    const cozyLabel =
      cozy && cozy.length > 0 && cozy[0].label ? cozy[0].label : undefined
    const emailValue =
      email && email.length > 0
        ? movePrimaryToHead(email).map(item => ({
            email: item.address,
            emailLabel: item.type
          }))
        : [undefined]
    const phoneValue =
      phone && phone.length > 0
        ? movePrimaryToHead(phone).map(item => ({
            phone: item.number,
            phoneLabel: item.type
          }))
        : [undefined]

    return {
      givenName: name ? name.givenName : undefined,
      familyName: name ? name.familyName : undefined,
      phone: phoneValue,
      email: emailValue,
      address: addressValue,
      cozy: cozyValue,
      cozyLabel: cozyLabel,
      company,
      birthday,
      note
    }
  }

  return initialFieldValues
}

export default contactToFormValues
