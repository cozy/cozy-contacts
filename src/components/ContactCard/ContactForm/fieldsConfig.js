export const fields = [
  {
    name: 'gender',
    icon: 'people',
    options: [
      {
        value: 'male',
        label: 'gender.man'
      },
      {
        value: 'female',
        label: 'gender.woman'
      }
    ]
  },
  {
    name: 'givenName',
    icon: null,
    type: 'text'
  },
  {
    name: 'additionalName',
    icon: null,
    type: 'text'
  },
  {
    name: 'familyName',
    icon: null,
    type: 'text'
  },
  {
    name: 'surname',
    icon: null,
    type: 'text'
  },
  {
    name: 'company',
    icon: 'company',
    type: 'text'
  },
  {
    name: 'jobTitle',
    icon: null,
    type: 'text'
  },
  {
    name: 'phone',
    icon: 'telephone',
    type: 'tel',
    label: {
      options: [
        {
          value: undefined,
          label: 'label.none'
        },
        {
          value: '{"type":"cell","label":"home"}',
          label: 'label.phone.cell-home'
        },
        {
          value: '{"type":"cell","label":"work"}',
          label: 'label.phone.cell-work'
        },
        {
          value: '{"type":"voice","label":"home"}',
          label: 'label.phone.voice-home'
        },
        {
          value: '{"type":"voice","label":"work"}',
          label: 'label.phone.voice-work'
        },
        {
          value: '{"type":"fax","label":"home"}',
          label: 'label.phone.fax-home'
        },
        {
          value: '{"type":"fax","label":"work"}',
          label: 'label.phone.fax-work'
        }
      ]
    },
    isArray: true
  },
  {
    name: 'email',
    icon: 'email',
    type: 'email',
    label: {
      options: [
        {
          value: undefined,
          label: 'label.none'
        },
        {
          value: '{"label":"home"}',
          label: 'label.home'
        },
        {
          value: '{"label":"work"}',
          label: 'label.work'
        }
      ]
    },
    isArray: true
  },
  {
    name: 'address',
    icon: 'location',
    type: 'button',
    subFields: [
      {
        name: 'street',
        icon: null,
        type: 'text'
      },
      {
        name: 'number',
        icon: null,
        type: 'text'
      },
      {
        name: 'building',
        icon: null,
        type: 'text'
      },
      {
        name: 'stairs',
        icon: null,
        type: 'text'
      },
      {
        name: 'floor',
        icon: null,
        type: 'text'
      },
      {
        name: 'apartment',
        icon: null,
        type: 'text'
      },
      {
        name: 'entrycode',
        icon: null,
        type: 'text'
      },
      {
        name: 'locality',
        icon: null,
        type: 'text'
      },
      {
        name: 'code',
        icon: null,
        type: 'text'
      },
      {
        name: 'city',
        icon: null,
        type: 'text'
      },
      {
        name: 'region',
        icon: null,
        type: 'text'
      },
      {
        name: 'country',
        icon: null,
        type: 'text'
      }
    ],
    label: {
      options: [
        {
          value: undefined,
          label: 'label.none'
        },
        {
          value: '{"label":"home"}',
          label: 'label.address.home'
        },
        {
          value: '{"label":"work"}',
          label: 'label.address.work'
        }
      ]
    },
    isArray: true
  },
  {
    name: 'cozy',
    icon: 'cloud',
    type: 'url',
    label: {
      options: [
        {
          value: undefined,
          label: 'label.none'
        },
        {
          value: '{"label":"home"}',
          label: 'label.home'
        },
        {
          value: '{"label":"work"}',
          label: 'label.work'
        }
      ]
    }
  },
  {
    name: 'birthday',
    icon: 'calendar',
    type: 'date',
    labelProps: { shrink: true }
  },
  {
    name: 'birthplace',
    icon: null,
    type: 'text'
  },
  {
    name: 'note',
    icon: 'comment',
    type: 'text',
    isMultiline: true
  }
]
