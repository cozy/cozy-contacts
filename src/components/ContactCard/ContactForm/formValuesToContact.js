import merge from 'lodash/merge'
import { updateIndexFullNameAndDisplayName } from '../../../helpers/contacts'

const formValuesToContact = (data, oldContact) => {
  const {
    givenName,
    familyName,
    phone,
    email,
    address,
    cozy,
    company,
    jobTitle,
    birthday,
    note
  } = data

  const contactWithFormValues = {
    name: {
      givenName,
      familyName
    },
    email: email
      .filter(val => val && val.email)
      .map(({ email, emailLabel }, index) => ({
        address: email,
        type: emailLabel,
        primary: index === 0
      })),
    address:
      address &&
      address
        .filter(val => val && val.address)
        .map(({ address, addressLabel }, index) => ({
          formattedAddress: address,
          type: addressLabel,
          primary: index === 0
        })),
    phone:
      phone &&
      phone
        .filter(val => val && val.phone)
        .map(({ phone, phoneLabel }, index) => ({
          number: phone,
          type: phoneLabel,
          primary: index === 0
        })),
    cozy: cozy
      ? [
          {
            url: cozy,
            label: data['cozyLabel'],
            primary: true
          }
        ]
      : [],
    company: company || '',
    jobTitle: jobTitle || '',
    birthday,
    note: note || '',
    // If we don't create the relationships field manually, cozy-client doesn't create it automatically when needed
    relationships: {
      groups: {
        data: []
      }
    },
    metadata: {
      version: 1,
      cozy: true
    }
  }

  return updateIndexFullNameAndDisplayName(
    merge({ ...oldContact }, contactWithFormValues)
  )
}

export default formValuesToContact
