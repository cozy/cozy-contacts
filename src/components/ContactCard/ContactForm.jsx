import React from "react";
import PropTypes from "prop-types";
import { Form } from "react-final-form";
import arrayMutators from "final-form-arrays";
import ContactFormField from "./ContactFormField";
import ContactFieldInput from "./ContactFieldInput";
import { Button } from "cozy-ui/react/Button";
import { translate } from "cozy-ui/react/I18n";

import IconEmail from "../../assets/icons/email.svg";
import IconPhone from "../../assets/icons/phone-number.svg";
import IconAddress from "../../assets/icons/location.svg";
import IconCozy from "../../assets/icons/cozy.svg";
import IconCompany from "../../assets/icons/company.svg";
import IconBirthday from "../../assets/icons/calendar.svg";
import IconNote from "../../assets/icons/comment.svg";

const fields = [
  {
    name: "givenName",
    icon: null,
    type: "text"
  },
  {
    name: "familyName",
    icon: null,
    type: "text",
    required: true
  },
  {
    name: "phone",
    icon: IconPhone,
    type: "tel",
    hasLabel: true,
    isArray: true
  },
  {
    name: "email",
    icon: IconEmail,
    type: "email",
    hasLabel: true,
    isArray: true
  },
  {
    name: "address",
    icon: IconAddress,
    type: "text",
    hasLabel: true,
    isArray: true
  },
  {
    name: "cozy",
    icon: IconCozy,
    type: "url",
    hasLabel: true
  },
  {
    name: "company",
    icon: IconCompany,
    type: "text"
  },
  {
    name: "birthday",
    icon: IconBirthday,
    type: "date"
  },
  {
    name: "note",
    icon: IconNote,
    type: "textarea"
  }
];

// initialize the form values, required so that array fields start with at least one editable field
const initialFieldValues = fields.reduce((initialValues, { name, isArray }) => {
  initialValues[name] = isArray ? [undefined] : undefined;
  return initialValues;
}, {});

class ContactForm extends React.Component {
  formDataToContact = data => {
    const {
      givenName,
      familyName,
      phone,
      email,
      address,
      cozy,
      company,
      birthday,
      note
    } = data;

    const fullName = (givenName || "") + " " + (familyName || "");

    const contact = {
      fullname: fullName.trim(),
      name: {
        givenName,
        familyName
      },
      email: email.filter(val => val).map(({ email, emailLabel }, index) => ({
        address: email,
        type: emailLabel,
        primary: index === 0
      })),
      address: address
        .filter(val => val)
        .map(({ address, addressLabel }, index) => ({
          formattedAddress: address,
          type: addressLabel,
          primary: index === 0
        })),
      phone: phone.filter(val => val).map(({ phone, phoneLabel }, index) => ({
        number: phone,
        type: phoneLabel,
        primary: index === 0
      })),
      groups: [],
      cozy: cozy
        ? [
            {
              url: cozy,
              label: data["cozyLabel"],
              primary: true
            }
          ]
        : undefined,
      company,
      birthday,
      note,
      metadata: {
        version: 1
      }
    };

    this.props.onSubmit(contact);
  };

  render() {
    const { onCancel, t } = this.props;
    return (
      <Form
        mutators={{ ...arrayMutators }}
        onSubmit={this.formDataToContact}
        initialValues={initialFieldValues}
        render={({ handleSubmit }) => (
          <div>
            <form onSubmit={handleSubmit} className="contact-form">
              <div className="contact-form__fields">
                {fields.map(
                  ({ name, icon, type, required, hasLabel, isArray }) => (
                    <ContactFormField
                      key={name}
                      name={name}
                      icon={icon}
                      label={t(`field.${name}`)}
                      isArray={isArray}
                      renderInput={inputName => (
                        <ContactFieldInput
                          name={inputName}
                          type={type}
                          placeholder={t(`placeholder.${name}`)}
                          required={required}
                          withLabel={hasLabel}
                          labelPlaceholder={t("placeholder.label")}
                        />
                      )}
                    />
                  )
                )}
              </div>

              <div className="contact-form__footer">
                <div>
                  <Button
                    type="button"
                    theme="secondary"
                    label={t("cancel")}
                    onClick={onCancel}
                  />
                  <Button type="submit" label={t("save")} />
                </div>
              </div>
            </form>
          </div>
        )}
      />
    );
  }
}

ContactForm.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired,
  t: PropTypes.func.isRequired
};

export default translate()(ContactForm);
