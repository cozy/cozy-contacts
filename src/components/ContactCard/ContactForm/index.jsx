import React from 'react'
import PropTypes from 'prop-types'
import { Form } from 'react-final-form'
import arrayMutators from 'final-form-arrays'
import { translate } from 'cozy-ui/transpiled/react/I18n'

import FieldsetTitle from '../../Common/FieldsetTitle'
import ContactFormField from './ContactFormField'
import ContactFieldInput from '../ContactFieldInput'
import { fullContactPropTypes } from '../../ContactPropTypes'
import contactToFormValues from './contactToFormValues'
import formValuesToContact from './formValuesToContact'
import { fields } from './fieldsConfig'

// this variable will be set in the form's render prop
// and used by the submit button in ContactFormModal
// to be able to trigger the submit from outside the form
// See react-final-form examples here: https://www.npmjs.com/package/react-final-form#external-submit
let _submitContactForm

function setSubmitContactForm(handleSubmit) {
  _submitContactForm = handleSubmit
}

export function getSubmitContactForm() {
  return _submitContactForm
}

const ContactForm = ({ contact, onSubmit, t }) => (
  <Form
    mutators={{ ...arrayMutators }}
    onSubmit={data => onSubmit(formValuesToContact(data, contact))}
    initialValues={contactToFormValues(contact, t)}
    render={({ handleSubmit }) => {
      setSubmitContactForm(handleSubmit)
      return (
        <div>
          <form onSubmit={handleSubmit} className="u-flex u-flex-column">
            <FieldsetTitle title={t('contact_info')} />
            {fields.map(
              ({
                name,
                icon,
                type,
                required,
                hasLabel,
                isArray,
                labelProps,
                isMultiline
              }) => (
                <ContactFormField
                  key={name}
                  name={name}
                  icon={icon}
                  label={t(`fields.${name}`)}
                  t={t}
                  isArray={isArray}
                  renderInput={(inputName, id) => (
                    <ContactFieldInput
                      id={id}
                      name={inputName}
                      type={type}
                      label={t(`fields.${name}`)}
                      required={required}
                      withLabel={hasLabel}
                      labelPlaceholder={t('fields.label')}
                      labelProps={labelProps}
                      isMultiline={isMultiline}
                    />
                  )}
                />
              )
            )}
          </form>
        </div>
      )
    }}
  />
)

ContactForm.propTypes = {
  contact: fullContactPropTypes,
  onSubmit: PropTypes.func.isRequired,
  t: PropTypes.func.isRequired
}

export default translate()(ContactForm)
