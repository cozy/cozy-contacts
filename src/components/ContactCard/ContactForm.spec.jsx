import React from "react";
import renderer from "react-test-renderer";
import { mount } from "enzyme";
import ContactForm from "./ContactForm";
import { I18n } from "cozy-ui/react/I18n";

describe("ContactForm", () => {
  it("should match snapshot", () => {
    const tree = renderer
      .create(
        <I18n lang="en" dictRequire={() => ""}>
          <ContactForm onSubmit={() => {}} onCancel={() => {}} />
        </I18n>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should submit a well formatted contact", () => {
    const onSubmit = contact => {
      expect(contact).toMatchSnapshot();
    };

    const form = mount(
      <I18n lang="en" dictRequire={() => ""}>
        <ContactForm onSubmit={onSubmit} onCancel={() => {}} />
      </I18n>
    );

    const fields = {
      givenName: "Jean-Claude",
      familyName: "Van Cozy",
      "phone[0].phone": "+33678987654",
      "email[0].email": "jcvc@cozy.cloud",
      "address[0].address": "18 rue des fleurs, Pecado",
      cozy: "https://jcvd.cozy.cloud",
      company: "Cozy CLoud",
      birthday: "31/12/2015",
      note: "Whatever."
    };

    Object.keys(fields).forEach(fieldName => {
      form
        .find(`Field[name='${fieldName}']`)
        .simulate("change", { target: { value: fields[fieldName] } });
    });

    form.find("form").simulate("submit");
  });

  it("should submit empty fields", () => {
    const onSubmit = contact => {
      expect(contact).toMatchSnapshot();
    };

    const form = mount(
      <I18n lang="en" dictRequire={() => ""}>
        <ContactForm onSubmit={onSubmit} onCancel={() => {}} />
      </I18n>
    );

    form.find("form").simulate("submit");
  });
});
