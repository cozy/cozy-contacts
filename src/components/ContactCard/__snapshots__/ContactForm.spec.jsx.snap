// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactForm should match snapshot 1`] = `
<div>
  <form
    className="contact-form"
    onSubmit={[Function]}
  >
    <div
      className="contact-form__fields"
    >
      <div
        className="contact-form__field"
      >
        <label
          className="contact-form__label"
        >
          field.givenName
        </label>
        <div
          className="contact-form__inputs-wrapper"
        >
          <div
            className="contact-form__input-wrapper"
          >
            <input
              className="contact-form__input"
              name="givenName"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="placeholder.givenName"
              required={false}
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        className="contact-form__field"
      >
        <label
          className="contact-form__label"
        >
          field.familyName
        </label>
        <div
          className="contact-form__inputs-wrapper"
        >
          <div
            className="contact-form__input-wrapper"
          >
            <input
              className="contact-form__input"
              name="familyName"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="placeholder.familyName"
              required={true}
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        className="contact-form__field"
      >
        <label
          className="contact-form__label"
        >
          <svg
            className="contact-form__icon icon"
            height="16"
            style={
              Object {
                "fill": "#95999D",
              }
            }
            width="16"
          >
            <use
              xlinkHref="#icon-0"
            />
          </svg>
          field.phone
        </label>
        <div
          className="contact-form__inputs-wrapper"
        >
          <div
            className="contact-form__meta-wrapper"
          >
            <div
              className="contact-form__input-wrapper"
            >
              <input
                className="contact-form__input"
                name="phone[0].phone"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="placeholder.phone"
                required={false}
                type="tel"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        className="contact-form__field"
      >
        <label
          className="contact-form__label"
        >
          <svg
            className="contact-form__icon icon"
            height="16"
            style={
              Object {
                "fill": "#95999D",
              }
            }
            width="16"
          >
            <use
              xlinkHref="#icon-0"
            />
          </svg>
          field.email
        </label>
        <div
          className="contact-form__inputs-wrapper"
        >
          <div
            className="contact-form__meta-wrapper"
          >
            <div
              className="contact-form__input-wrapper"
            >
              <input
                className="contact-form__input"
                name="email[0].email"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="placeholder.email"
                required={false}
                type="email"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        className="contact-form__field"
      >
        <label
          className="contact-form__label"
        >
          <svg
            className="contact-form__icon icon"
            height="16"
            style={
              Object {
                "fill": "#95999D",
              }
            }
            width="16"
          >
            <use
              xlinkHref="#icon-0"
            />
          </svg>
          field.address
        </label>
        <div
          className="contact-form__inputs-wrapper"
        >
          <div
            className="contact-form__meta-wrapper"
          >
            <div
              className="contact-form__input-wrapper"
            >
              <input
                className="contact-form__input"
                name="address[0].address"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="placeholder.address"
                required={false}
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        className="contact-form__field"
      >
        <label
          className="contact-form__label"
        >
          <svg
            className="contact-form__icon icon"
            height="16"
            style={
              Object {
                "fill": "#95999D",
              }
            }
            width="16"
          >
            <use
              xlinkHref="#icon-0"
            />
          </svg>
          field.cozy
        </label>
        <div
          className="contact-form__inputs-wrapper"
        >
          <div
            className="contact-form__input-wrapper"
          >
            <input
              className="contact-form__input"
              name="cozy"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="placeholder.cozy"
              required={false}
              type="url"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        className="contact-form__field"
      >
        <label
          className="contact-form__label"
        >
          <svg
            className="contact-form__icon icon"
            height="16"
            style={
              Object {
                "fill": "#95999D",
              }
            }
            width="16"
          >
            <use
              xlinkHref="#icon-0"
            />
          </svg>
          field.company
        </label>
        <div
          className="contact-form__inputs-wrapper"
        >
          <div
            className="contact-form__input-wrapper"
          >
            <input
              className="contact-form__input"
              name="company"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="placeholder.company"
              required={false}
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        className="contact-form__field"
      >
        <label
          className="contact-form__label"
        >
          <svg
            className="contact-form__icon icon"
            height="16"
            style={
              Object {
                "fill": "#95999D",
              }
            }
            width="16"
          >
            <use
              xlinkHref="#icon-0"
            />
          </svg>
          field.birthday
        </label>
        <div
          className="contact-form__inputs-wrapper"
        >
          <div
            className="contact-form__input-wrapper"
          >
            <input
              className="contact-form__input"
              name="birthday"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="placeholder.birthday"
              required={false}
              type="date"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        className="contact-form__field"
      >
        <label
          className="contact-form__label"
        >
          <svg
            className="contact-form__icon icon"
            height="16"
            style={
              Object {
                "fill": "#95999D",
              }
            }
            width="16"
          >
            <use
              xlinkHref="#icon-0"
            />
          </svg>
          field.note
        </label>
        <div
          className="contact-form__inputs-wrapper"
        >
          <div
            className="contact-form__input-wrapper"
          >
            <textarea
              className="contact-form__input"
              name="note"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="placeholder.note"
              required={false}
              type="textarea"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="contact-form__footer"
    >
      <div>
        <button
          className="c-btn c-btn--secondary"
          onClick={[Function]}
          role="button"
          type="button"
        >
          <span>
            <span>
              cancel
            </span>
          </span>
        </button>
        <button
          className="c-btn"
          role="button"
          type="submit"
        >
          <span>
            <span>
              save
            </span>
          </span>
        </button>
      </div>
    </div>
  </form>
</div>
`;

exports[`ContactForm should submit a well formatted contact 1`] = `
Object {
  "address": Array [
    Object {
      "formattedAddress": "18 rue des fleurs, Pecado",
      "primary": true,
      "type": undefined,
    },
  ],
  "birthday": "31/12/2015",
  "company": "Cozy CLoud",
  "cozy": Array [
    Object {
      "label": undefined,
      "primary": true,
      "url": "https://jcvd.cozy.cloud",
    },
  ],
  "email": Array [
    Object {
      "address": "jcvc@cozy.cloud",
      "primary": true,
      "type": undefined,
    },
  ],
  "fullname": "Jean-Claude Van Cozy",
  "groups": Array [],
  "metadata": Object {
    "cozy": true,
    "version": 1,
  },
  "name": Object {
    "familyName": "Van Cozy",
    "givenName": "Jean-Claude",
  },
  "note": "Whatever.",
  "phone": Array [
    Object {
      "number": "+33678987654",
      "primary": true,
      "type": undefined,
    },
  ],
}
`;

exports[`ContactForm should submit empty fields 1`] = `
Object {
  "address": Array [],
  "birthday": undefined,
  "company": undefined,
  "cozy": undefined,
  "email": Array [],
  "fullname": "",
  "groups": Array [],
  "metadata": Object {
    "cozy": true,
    "version": 1,
  },
  "name": Object {
    "familyName": undefined,
    "givenName": undefined,
  },
  "note": undefined,
  "phone": Array [],
}
`;
